Upgrade cmdliner fork to 1.1.1

emillon opened the pull request:
We still need a fork to support `alias` but this brings the upstream part to 1.1.1. The main addition is builtin support of groups through the `Cmdliner.Cmd` API.

Benefits include:

- we get closer to upstream `cmdliner`
- help pages like `dune ocaml --help` are now more useful

This commit contains several things:

- an update of the vendored copy (to be reviewed separately),
- a port of `bin/` to the `Cmdliner.Cmd` API,
- test updates, mostly typographic.

emillon commented:
I'll put comments with places to discuss. The PR on cmdliner is https://github.com/ocaml-dune/cmdliner/pull/1.

emillon commented on bin/import.ml:
```diff
@@ -161,9 +160,8 @@ let restore_cwd_and_execve (common : Common.t) prog argv env =
-let command_alias cmd name =
-  let term, info = cmd in
-  let orig = Term.name info in
+let command_alias cmd term name =
```
There's no public way to get `term` from `cmd`, so I passed it and exposed the term in `Build_cmd`.

emillon commented on bin/main.ml:
```diff
@@ -45,52 +44,48 @@ let common_commands_synopsis =
 let () =
   Colors.setup_err_formatter_colors ();
   try
-    match Term.Group.eval default all ~catch:false with
```
I tested all return values of `eval_value` and it looks equivalent but I'm not completely confident about this

emillon commented on test/blackbox-tests/test-cases/cmdline/profile.t:
```diff
@@ -1,9 +1,9 @@
 # Bug #4632
 
   $ dune build -p
-  dune build: option `-p' needs an argument
```
error messages do not output the command name now. I haven't tried to see if this is something that we can fix but this looks like a regression with how the messages look

rgrinberg replied:
Did you ask upstream about his?

emillon replied:
Good point. I started working a patch to see if this can fit in cmdliner.

emillon replied:
I patched locally and will submit the PR upstream.

emillon commented on bin/init.ml:
```diff
@@ -249,4 +249,4 @@ let term =
   print_completion kind name
 
-let command = (term, info)
+let command = Cmd.v info term
```
This file (and probably others) mention "when cmdliner supports group, update it". we should create corresponding issues

emillon replied:
Done in #6161

christinerose commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -251,24 +251,23 @@ Comments in dune files are preserved
 Will not create components with invalid names
 
   $ dune init lib invalid-component-name ./_test_lib
```
```suggestion
  $ dune init lib invalid-component-name ./_test_lib
```
Should these be in code block or monospace. I think it's up to you, as it's clearly something for the command line.

emillon replied:
This is cram syntax, this is interpreted.

christinerose commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -251,24 +251,23 @@ Comments in dune files are preserved
+        Library names must be non-empty and composed only of the
+        following
+        characters: 'A'..'Z', 'a'..'z', '_' or '0'..'9'.
+  Usage: dune init [OPTION]… COMPONENT NAME [PATH]
```
```suggestion
  Usage: `dune init` [OPTION]… COMPONENT NAME [PATH]
```

christinerose commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -251,24 +251,23 @@ Comments in dune files are preserved
-             `executable', `library', `project' or `test'
-  Usage: dune init [OPTION]... COMPONENT NAME [PATH]
-  Try `dune init --help' or `dune --help' for more information.
+  dune: COMPONENT argument: invalid value 'foo', expected one of 'executable',
```
```suggestion
 `dune`: COMPONENT argument: invalid value 'foo', expected one of 'executable',
```

Same as previous comments, although I think it's pretty clear it's the command without the monospace. It would just be fore consistency.

There are several others below, but I won't make this suggestion on all of them until you decide which you prefer. Then I'm happy to go through it again. :)

christinerose commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -251,24 +251,23 @@ Comments in dune files are preserved
-  Try `dune init --help' or `dune --help' for more information.
+  dune: COMPONENT argument: invalid value 'foo', expected one of 'executable',
+        'library', 'project' or 'test'
+  Usage: dune init [OPTION]… COMPONENT NAME [PATH]
```
```suggestion
  Usage: `dune init` [OPTION]… COMPONENT NAME [PATH]
```

Same as above...

christinerose commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -251,24 +251,23 @@ Comments in dune files are preserved
+  Try 'dune init --help' or 'dune --help' for more information.
   [1]
 
 Will fail and inform user when an invalid option is given to a component
```
```suggestion
This command will fail and inform user when an invalid option is given to a component
```

emillon replied:
This is better but unrelated to this PR so I'd rather leave it off the diff.

christinerose commented on vendor/cmdliner/LICENSE.md:
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2011 Daniel C. Bünzli
+Copyright (c) 2011 The cmdliner programmers
```
```suggestion
Copyright (c) 2011 The `Cmdliner` Programmers
```

Since `Cmdliner` is capitalised in the code, perhaps it would be good to be capitalized and in monospace here. If you do, perhaps capitalise Programmers too.

If you choose to do this, I will go through and suggest on all of them, so you can just accept suggestion. Let me know! :)

emillon replied:
Everything under `vendor/` is pulled from third party code so we can't change it.

emillon commented:
Actually, regarding displaying command names or not, this is a feature that we introduced in our fork to implement groups in #3936. So it's not a regression in cmdliner. I'd rather keep the vanilla cmdliner code rather than patch it to keep the behaviour we introduced.

To clarify, it means that the first line of the message if

    dune: TARGET... arguments: unclosed parenthesis at end of input

rather than

    dune build: TARGET... arguments: unclosed parenthesis at end of input

but it's ok since the following lines are 

    Usage: dune build [OPTION]... [TARGET]...
    Try `dune build --help' or `dune --help' for more information.

Alizter commented:
I'm in favor of dropping the command name patch since it doesn't hinder the error message and is easier to maintain.

emillon commented:
Thanks! I've closed the cmdliner issue, will merge the ocaml-dune/cmdliner PR, and take care of the rest here.
